File: app/src/main/java/cn/devcxl/photosync/ptp/manager/SyncDeviceManager.java
package cn.devcxl.photosync.ptp.manager;

import android.hardware.usb.UsbDevice;
import android.util.Log;

import java.util.ArrayList;
import java.util.Date;
import java.util.List;

import cn.devcxl.photosync.App;
import cn.devcxl.photosync.data.AppDatabase;
import cn.devcxl.photosync.data.SyncDeviceDao;
import cn.devcxl.photosync.data.entity.SyncDevice;
import cn.devcxl.photosync.ptp.detect.CameraDetector;

/**
 * !!! Not thread safe !!!
 * 理论上我们的程序应该运行在单一线程，所以为了性能考虑，没有加锁
 *
 * @author rainx
 * @date 2017/5/27
 */
public class SyncDeviceManager {
    public static final String TAG = "SyncDeviceManager";

    private UsbDevice device;
    private final SyncDeviceDao dao;

    public SyncDeviceManager(UsbDevice device) {
        this.device = device;
        this.dao = AppDatabase.getInstance(App.Companion.get()).syncDeviceDao();
    }


    /**
     * 更新同步设备的基本信息， 如果设备信息已经存在，则只更新时间
     *
     * @return true: if first time create
     *          null: if already exists
     *
     */
    public SyncDevice updateDeviceInfo()  {
        String uuid = getUUIDFromDevice(device);
        SyncDevice syncDevice = getSyncDevice(device);

        long now = new Date().getTime();
        if (syncDevice == null) {
            String serial;
            try {
                serial = device.getSerialNumber();
            } catch (SecurityException se) {
                Log.w(TAG, "No permission to read serial number when updating device info; using empty", se);
                serial = "";
            }
            syncDevice = new SyncDevice(
                    uuid,
                    device.getDeviceName(),
                    device.getProductName(),
                    device.getManufacturerName(),
                    device.getVendorId(),
                    device.getDeviceId(),
                    device.getProductId(),
                    serial,
                    //device.getVersion() require android version 23;
                    device.getVersion(),
                    "",
                    false,
                    null,
                    now,
                    now
            );
            dao.insert(syncDevice);
            Log.d(TAG, "device recording" + syncDevice);
        } else {
            // 设备已经记录了
            syncDevice.setUpdatedAt(now);
            dao.update(syncDevice);
            Log.d(TAG, "device already recorded");
        }
        return syncDevice;
    }

    /**
     * 在当前的手机上清除拍照设备信息
     *
     * @return 是否执行了删除操作
     */
    public boolean removeDevice() {
        SyncDevice syncDevice = getSyncDevice(device);
        if (syncDevice != null) {
            return dao.delete(syncDevice) > 0;
        } else {
            return false;
        }
    }

    /**
     * 开始同步
     */
    public void startSync() {
        SyncDevice syncDevice = getSyncDevice(device);

        if (syncDevice.getSyncAt() == null || syncDevice.getSyncAt() == 0) {
            syncDevice.setSyncAt(new Date().getTime());
        }

        syncDevice.setUpdatedAt(new Date().getTime());
        syncDevice.setSyncing(true);
        dao.update(syncDevice);
    }

    /**
     * 停止同步
     */
    public void stopSync() {
        SyncDevice syncDevice = getSyncDevice(device);

        syncDevice.setUpdatedAt(new Date().getTime());
        syncDevice.setSyncing(false);
        dao.update(syncDevice);
    }


    /**
     * 更新id列表
     * @param ids
     */
    public void updateIdList(List<Integer> ids) {
        StringBuffer strIdListBuffer = new StringBuffer();
        if (ids != null) {
            int i = 1;
            for (Integer id : ids){
                strIdListBuffer.append(id);
                if (i++ != ids.size()) {
                    strIdListBuffer.append(",");
                }
            }
        }
        String strIdList = strIdListBuffer.toString();

        SyncDevice syncDevice = getSyncDevice(device);
        syncDevice.setSyncIdList(strIdList);
        syncDevice.setSyncAt(new Date().getTime());
        syncDevice.setUpdatedAt(new Date().getTime());
        dao.update(syncDevice);
    }

    /**
     * 获取id列表
     * @return
     */
    public List<Integer> getIdList() {
        SyncDevice syncDevice = getSyncDevice(device);
        return getIdList(syncDevice);
    }

    /***
     * 根据syncDevice获取id列表
     */
    public List<Integer> getIdList(SyncDevice syncDevice) {
        String strIdList = syncDevice.getSyncIdList();
        List<Integer> idList = new ArrayList<>();
        if (strIdList == null || strIdList.equals("")) {
            return idList;
        }

        String[] strIdArr = strIdList.split(",");
        for(String id : strIdArr) {
            idList.add(Integer.valueOf(id));
        }
        return idList;
    }

    /**
     * 获取数据库中的同步设备ORM
     * @param device
     * @return
     */
    public SyncDevice getSyncDevice(UsbDevice device) {
        String uuid = getUUIDFromDevice(device);
        Log.d(TAG, "获取数据库中的同步设备ORM:" + uuid );
        SyncDevice found = dao.getByUuid(uuid);
        if (found != null) {
            return found;
        } else {
            Log.d(TAG, "没有找到对应的设备");
            return null;
        }
    }


    public UsbDevice getDevice() {
        return device;
    }

    public void setDevice(UsbDevice device) {
        this.device = device;
    }

    /**
     * 获取所有同步设备信息
     * @return
     */
    public List<SyncDevice> getAllSyncDevices() {
        return dao.getAll();
    }

    private String getUUIDFromDevice(UsbDevice device) {
        CameraDetector cd = new CameraDetector(device);
        String uuid = cd.getDeviceUniqName();
        return uuid;
    }



}

File: app/src/main/java/cn/devcxl/photosync/ptp/manager/SyncDeviceManager.kt
package cn.devcxl.photosync.ptp.manager

import android.hardware.usb.UsbDevice
import cn.devcxl.photosync.App
import cn.devcxl.photosync.data.AppDatabase
import cn.devcxl.photosync.data.entity.SyncDevice
import cn.devcxl.photosync.ptp.detect.CameraDetector
import timber.log.Timber
import java.util.Date

/**
 * !!! Not thread safe !!!
 * @author devcxl
 */
class SyncDeviceManager(private var device: UsbDevice) {

    private val dao = AppDatabase.getInstance(App.get()!!).syncDeviceDao()

    fun updateDeviceInfo(): SyncDevice? {
        val uuid = getUUIDFromDevice(device)
        var syncDevice = getSyncDevice(device)
        val now = Date().time
        if (syncDevice == null) {
            val serial = try {
                device.serialNumber
            } catch (se: SecurityException) {
                Timber.tag(TAG).w(se, "No permission to read serial number when updating device info; using empty")
                ""
            }
            syncDevice = SyncDevice(
                deviceUUID = uuid,
                deviceName = device.deviceName,
                productName = device.productName,
                manufacturerName = device.manufacturerName,
                vendorId = device.vendorId,
                deviceId = device.deviceId,
                productId = device.productId,
                serialNumber = serial,
                version = device.version,
                syncIdList = "",
                isSyncing = false,
                syncAt = null,
                updatedAt = now,
                createdAt = now
            )
            dao.insert(syncDevice)
            Timber.tag(TAG).d("device recording $syncDevice")
        } else {
            syncDevice.updatedAt = now
            dao.update(syncDevice)
            Timber.tag(TAG).d("device already recorded")
        }
        return syncDevice
    }

    fun removeDevice(): Boolean {
        val syncDevice = getSyncDevice(device)
        return if (syncDevice != null) {
            dao.delete(syncDevice) > 0
        } else {
            false
        }
    }

    fun startSync() {
        val syncDevice = getSyncDevice(device) ?: return
        if (syncDevice.syncAt == null || syncDevice.syncAt == 0L) {
            syncDevice.syncAt = Date().time
        }
        syncDevice.updatedAt = Date().time
        syncDevice.isSyncing = true
        dao.update(syncDevice)
    }

    fun stopSync() {
        val syncDevice = getSyncDevice(device) ?: return
        syncDevice.updatedAt = Date().time
        syncDevice.isSyncing = false
        dao.update(syncDevice)
    }

    fun updateIdList(ids: List<Int>?) {
        val sb = StringBuilder()
        if (ids != null) {
            var i = 1
            for (id in ids) {
                sb.append(id)
                if (i++ != ids.size) {
                    sb.append(",")
                }
            }
        }
        val strIdList = sb.toString()
        val syncDevice = getSyncDevice(device) ?: return
        syncDevice.syncIdList = strIdList
        syncDevice.syncAt = Date().time
        syncDevice.updatedAt = Date().time
        dao.update(syncDevice)
    }

    fun getIdList(): List<Int> {
        val syncDevice = getSyncDevice(device) ?: return emptyList()
        return getIdList(syncDevice)
    }

    fun getIdList(syncDevice: SyncDevice): List<Int> {
        val strIdList = syncDevice.syncIdList ?: ""
        if (strIdList.isEmpty()) {
            return emptyList()
        }
        return strIdList.split(",").map { it.toInt() }
    }

    fun getSyncDevice(device: UsbDevice): SyncDevice? {
        val uuid = getUUIDFromDevice(device)
        Timber.tag(TAG).d("获取数据库中的同步设备ORM:$uuid")
        val found = dao.getByUuid(uuid)
        if (found != null) {
            return found
        } else {
            Timber.tag(TAG).d("没有找到对应的设备")
            return null
        }
    }

    fun getDevice(): UsbDevice = device

    fun setDevice(device: UsbDevice) {
        this.device = device
    }

    fun getAllSyncDevices(): List<SyncDevice> = dao.getAll()

    private fun getUUIDFromDevice(device: UsbDevice): String {
        val cd = CameraDetector(device)
        return cd.getDeviceUniqName()
    }

    companion object {
        const val TAG = "SyncDeviceManager"
    }
}
